test/distributed/test_ddp_under_dist_autograd.py

```diff
@@ -0,0 +1,474 @@
+        if self.rank == MASTER_RANK:
+            self._master_process(ddp_mode)
+        elif self.rank == REMOTE_WORKER_RANK:
+            self._remote_worker_process()
+        elif self.rank in TRAINER_RANKS:
+            self._trainer_process(self.rank)
+        else:
+            raise RuntimeError(f"Unknow process rank: {self.rank}")
+
+    @requires_gloo()
+    @dist_init
+    def test_backward_no_ddp(self):
+        self._do_test(DdpMode.NONE)
+
+    @requires_gloo()
+    @dist_init
+    def test_backward_ddp_outside(self):
+        self._do_test(DdpMode.OUTSIDE)
+
+    @requires_gloo()
+    @dist_init
+    def test_backward_ddp_inside(self):
+        self._do_test(DdpMode.INSIDE)
+
+
+@unittest.skipIf(
+    TEST_WITH_ASAN,
+    "Skip ASAN as torch + multiprocessing spawn have known issues",
+)
+class TestDdpComparison(MultiProcessTestCase):
+    rpc_backend = rpc.backend_registry.BackendType.PROCESS_GROUP
+    rpc_backend_options = None
+
+    @property
+    def world_size(self) -> int:
+        return NUM_TRAINERS
+
+    def trainer_name(self, rank):
+        # The name has to be consistent with that in 'dist_init' decorator.
+        return f"worker{rank}"
+
+    def setUp(self):
+        super(TestDdpComparison, self).setUp()
+
+        os.environ["MASTER_ADDR"] = str(MASTER_ADDR)
+        os.environ["MASTER_PORT"] = str(MASTER_PORT)
+        self._spawn_processes()
+
+    def tearDown(self):
+        super(TestDdpComparison, self).tearDown()
+
+    @requires_gloo()
+    @dist_init
+    def test_ddp_comparison(self):
+        gLogger.info(f"Running trainer rank: {self.rank}")
+        process_group_for_ddp = dist_c10d.new_group(ranks=TRAINER_RANKS)
+        net = nn.Linear(2, 3)
+        ddp_net = DistributedDataParallel(
+            net, process_group=process_group_for_ddp
+        )
+        inputs = torch.rand((3, 2))
+
+        # Use local autograd. The gradients will be in each variable's '.grad'.
+        loss = ddp_net(inputs).norm()
+        loss.backward()
+
+        # Use distributed autograd. The gradients will be in RPC context map.
+        grads_dict = {}
+        with dist_autograd.context() as context_id:
+            loss = ddp_net(inputs).norm()
+            dist_autograd.backward(context_id, [loss])
+            grads_dict = dist_autograd.get_gradients(context_id)
+        gLogger.info(f"Trainer #{self.rank} got grad dict: {grads_dict}")
+        # The gradients should be the same
+        for param in net.parameters():
+            self.assertTrue(
+                param in grads_dict,
+                f"Param {param} is not in dist_auto grad dict {grads_dict}",
+            )
+            grad_diff = grads_dict[param] - param.grad
+            self.assertAlmostEqual(
+                0.0,
+                float(grad_diff.norm()),
+                msg=f"The grads for param {param} are different under local "
+                f"and dist autograd: {param.grad} \n---\n {grads_dict[param]}",
+            )
```

Comment:
Why don't we just compare the  two Tensors instead of having a diff and a norm? We override assertEqual to support Tensor comparisons: https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_utils.py#L986.

Reply:
Good point. I didn't know that. Will make the change.